src/tree/builder/TreeAdapter.js

```diff
@@ -0,0 +1,449 @@
+	 * (e.g. "http://www.w3.org/1999/xhtml" for HTML)
+	 *
+	 * This is used by `parse5` to differentiate between parsing
+	 * HTML, SVG and other XML schema types.
+	 *
+	 * @param {Node} node The node to get the namespace URI from.
+	 *
+	 * @returns {string} The namespace URI of this node.
+	 */
+	getNamespaceURI( node ) {
+		return node.namespace;
+	}
+
+	/**
+	 * Returns this node's parent node.
+	 *
+	 * @param {Node} node	The node from which to retrieve the parent.
+	 *
+	 * @returns {Node} The parent of this node.
+	 */
+	getParentNode( node ) {
+		return node.parent;
+	}
+
+	/**
+	 * Returns the children of the given node.
+	 *
+	 * If the node does not have any children and cannot get any (e.g. Heading, FormattingElement)
+	 * this function returns an empty list.
+	 *
+	 * @param {Node} node	The node to get the children from.
+	 *
+	 * @returns {Node[]} The children of the given node.
+	 */
+	getChildNodes( node ) {
+		/*
+		  Some node types do not have children (like Paragraph and Heading),
+		  but parse5 always expects a node to have children.
+		 */
+		return node.children || [];
+	}
+
+	/**
+	 * Gets a parent's first child.
+	 *
+	 * @see https://en.wikipedia.org/wiki/Rumpelstiltskin
+	 *
+	 * @param {Node} node	The node to get its first child from.
+	 *
+	 * @returns {Node[]|null} The node's first child or null, if this node cannot get any children.
+	 */
+	getFirstChild( node ) {
+		return node.children[ 0 ] || null;
+	}
+
+	// Node source location.
+
+	/**
+	 * Sets the node's location as found in the source code.
+	 *
+	 * More often than not this is not the entire info
+	 * and misses the location of the end tag...
+	 *
+	 * We still need to add it, since `parse5` appends the end tag position
+	 * to this object somewhere during parsing.
+	 *
+	 * @param {Node} node	The node to set its location.
+	 * @param {Location} location the node's location in the source code.
+	 *
+	 * @returns {void}
+	 */
+	setNodeSourceCodeLocation( node, location ) {
+		if ( ! node ) {
+			return;
+		}
+		node.location = location;
+	}
+
+	/**
+	 * Gets the node's source code location.
+	 *
+	 * @param {Node} node	The node to get its source code location from.
+	 * @returns {Location|void} The node's source code location.
+	 */
+	getNodeSourceCodeLocation( node ) {
+		if ( ! node ) {
+			return;
+		}
+		return node.location;
+	}
+
+	// Private utility methods.
+
+	/**
+	 * Finds the most recent ancestor (parent of parent of ... ) of this node that returns true
+	 * on the given predicate.
+	 *
+	 * @param {Node|FormattingElement} element  The node to find the ancestor of.
+	 * @param {predicate} predicate	             The predicate to check the ancestors on.
+	 *
+	 * @returns {Node|null} The most recent ancestor that returns true on the given predicate.
+	 *
+	 * @private
+	 */
+	static _findAncestor( element, predicate ) {
+		let parent = element.parent;
+		while ( ! predicate( parent ) && parent.parent !== null ) {
```

Comment:
I think this was a brainfart on my part.
I changed it so it checks if `parent` is `null` instead of `parent.parent`, which should make more sense.

I also added an extra inline comment explaining the workings of the loop. In my opinion adding an extra line of code would not make it more readable in this case.